### PR TITLE
To get the [request object] in the [cart hook] and to be able to call the cart hook [with parameters]...

### PR DESCRIPTION
**Pull request by codixio**
... the request object is needed. E.g. to manipulate the line items in the cart. In a custom hook I do not have the services.cart object available.

### 1. Why is this change necessary?

Extends the cart hook with the possibility to read the request object to pass parameters to the cart hook.

### 2. What does this change do, exactly?

Adds RequestFacadeFactory::class in the getServiceIds array.
Commented out "\assert($request !== null);" in the RequestFacadeFactory to allow Request being null.
Changed constructor in RequestFacade to allow Request being null.
Changed all methods of RequestFacade to allow returning null instead of throwing an exeption in that case (more compatible/non-breaking for everyone using the request facade before ).

### 3. Describe each step to reproduce the issue or behaviour.

services.request.query is not available before, but it is with this addition now

{% if services.request.query.foo === 'bar' %}
    {# clear the cart #}
{% endif %}

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by codixio** [comment]
@dneustadt I made the changes to the cart hook and request facade

I'm not sure why the bot "Danger / pr (pull_request_target)" failed with "You should not use the same commit message multiple times" but this should not influence the pull request I think.

File: src/Core/Framework/Routing/Facade/RequestFacade.php [hunk 1]
```diff
@@ -128,7 +128,7 @@ public function request(): array
      */
     public function headers(): array
```
Comment (dneustadt) [inline src/Core/Framework/Routing/Facade/RequestFacade.php]:
```suggestion
    public function headers(): ?array
    {
        if (!$this-request) {
            return null;
        }
        
        $headers = array_change_key_case($this->request->headers->all());
```

File: src/Core/Framework/Routing/Facade/RequestFacade.php [hunk 2]
```diff
@@ -76,7 +76,7 @@ public function scheme(): string
      */
     public function method(): string
```
Comment (dneustadt) [inline src/Core/Framework/Routing/Facade/RequestFacade.php]:
```suggestion
    public function method(): ?string
```

Update all the return types. The phpdoc comments as welll: `@return ?string`

File: src/Core/Framework/Routing/Facade/RequestFacade.php (hunk 1 above)
Comment (codixio) [reply src/Core/Framework/Routing/Facade/RequestFacade.php]: Hi David, I changed that in my last commit. Thank you.

File: src/Core/Framework/Routing/Facade/RequestFacade.php (hunk 2 above)
Comment (codixio) [reply src/Core/Framework/Routing/Facade/RequestFacade.php]: Hi David, I changed all that in my last commit as well. Thank you for letting me know.

File: src/Core/Framework/Routing/Facade/RequestFacade.php [hunk 3]
```diff
@@ -126,8 +126,12 @@ public function request(): array
      *
      * @return array<string, array<int, string|null>|string|null> request headers
```
Comment (dneustadt) [inline src/Core/Framework/Routing/Facade/RequestFacade.php]:
```suggestion
     * @return array<string, array<int, string|null>|string|null>|null request headers
```

File: src/Core/Framework/Routing/Facade/RequestFacadeFactory.php [hunk 4]
```diff
@@ -21,7 +21,8 @@ public function __construct(private readonly RequestStack $requestStack)
     public function factory(Hook $hook, Script $script): RequestFacade
     {
         $request = $this->requestStack->getMainRequest();
-        \assert($request !== null);
+        // assert disabled: null is now allowed
+        //\assert($request !== null);
```
Comment (dneustadt) [inline src/Core/Framework/Routing/Facade/RequestFacadeFactory.php]: Not necessary to leave the former line commented out. Both comments can be removed.

File: src/Core/Framework/Routing/Facade/RequestFacade.php [hunk 5]
```diff
@@ -126,8 +126,12 @@ public function request(): array
...
      */
-    public function headers(): array
+    public function headers(): ?array
     {
+        if (!$this-request) {
```
Comment (dneustadt) [inline src/Core/Framework/Routing/Facade/RequestFacade.php]:
```suggestion
        if (!$this->request) {
```

File: src/Core/Framework/Routing/Facade/RequestFacade.php (hunk 3 above)
Comment (codixio) [reply src/Core/Framework/Routing/Facade/RequestFacade.php]: I changed and committed that.

File: src/Core/Framework/Routing/Facade/RequestFacadeFactory.php (hunk 4 above)
Comment (codixio) [reply src/Core/Framework/Routing/Facade/RequestFacadeFactory.php]: I changed and committed that.

File: src/Core/Framework/Routing/Facade/RequestFacade.php (hunk 5 above)
Comment (codixio) [reply src/Core/Framework/Routing/Facade/RequestFacade.php]: I changed and committed that.

**Comment by OliverSkroblin** [comment]
Hey, I discussed your PR with the colleagues, and access to the Request object within the Cart is not desired by us.

Establishing an (optional) dependency between Cart Processor and Request is undesirable, and it's something we've consistently avoided in recent years.

To transport data between Request and Cart via apps, it would be necessary to use custom controllers, LineItems, or other hooks beforehand to transmit the data to the shopping cart.

**Comment by OliverSkroblin** [comment]
Due to my last comment I'm closing this PR.
If you would like to discuss further implementation options and solutions, please feel free to create a new discussion.

**Pull request closed without merging**